src/backend/TxCache.cpp

```diff
@@ -0,0 +1,105 @@
+#include <backend/TxCache.h>
+namespace Backend {
+
+uint32_t
+TxCache::latestLedgerSequence() const
+{
+    std::shared_lock lck{mtx_};
+    return latestSeq_;
+}
+
+// adds an entire ledger's worth of transactions to the cache. Evicts oldest
+// ledger.
+void
+TxCache::update(
+    std::vector<ripple::uint256> const& hashes,
+    std::vector<TransactionAndMetadata> const& transactions,
+    uint32_t seq)
+{
+    std::unique_lock lck{mtx_};
+    cache_[tail_].clear();
+
+    assert(seq == latestSeq_ + 1 || latestSeq_ == 0);
+    latestSeq_ = seq;
+    for (std::size_t i = 0; i < hashes.size(); i++)
+    {
+        cache_[tail_][hashes[i]] = transactions[i];
+    }
+    tail_ = (tail_ + 1) % NUM_LEDGERS_CACHED;
```

Comment:
So I actually think this stores duplicates when not in readonly mode. 

We call txCache().update() twice, first in `updateTransactions`, then again in `publishLedger()`, meaning we increment tail_ twice for each ledger, storing a duplicate. I don't think this affects correctness, It just means it only stores `NUM_LEDGERS_CACHED / 2` ledgers worth of transactions in the cache.

We can fix this by only updating the cache when publishing a ledger, see comments below. 

I haven't run the code yet, so try to observe this behavior by adding some `trace` log statements before going to fix it.